File: backend/finacialsim_saas/schemas/clients.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel

from finacialsim_saas.schemas.types import DecimalStr


class ClientIn(BaseModel):
    nome: str
    cpf_cnpj: str
    tipo: str  # "pf" | "pj"
    rg: str | None = None
    data_nasc: date | None = None
    profissao: str | None = None
    renda: DecimalStr | None = None
    telefone: str | None = None
    email: str | None = None
    endereco_json: dict | None = None
    observacoes: str | None = None


class ClientOut(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    nome: str
    cpf_cnpj: str
    tipo: str
    rg: str | None
    data_nasc: date | None
    profissao: str | None
    renda: DecimalStr | None
    telefone: str | None
    email: str | None
    endereco_json: dict | None
    observacoes: str | None
    is_active: bool
    criado_por: uuid.UUID
    criado_em: datetime
    atualizado_em: datetime

    class Config:
        from_attributes = True


class ClientListItem(BaseModel):
    id: uuid.UUID
    nome: str
    cpf_cnpj: str
    tipo: str
    telefone: str | None
    email: str | None
    is_active: bool
    criado_em: datetime

    class Config:
        from_attributes = True


class ClientListPage(BaseModel):
    items: list[ClientListItem]
    next_cursor: str | None

File: frontend/src/lib/clients.ts
import { api } from "./api";

export interface ClientOut {
  id: string;
  tenant_id: string;
  nome: string;
  cpf_cnpj: string;
  tipo: "pf" | "pj";
  rg: string | null;
  data_nasc: string | null;
  profissao: string | null;
  renda: string | null;
  telefone: string | null;
  email: string | null;
  endereco_json: Record<string, string> | null;
  observacoes: string | null;
  is_active: boolean;
  criado_por: string;
  criado_em: string;
  atualizado_em: string;
}

export interface ClientListPage {
  items: ClientOut[];
  next_cursor: string | null;
}

export interface ClientIn {
  nome: string;
  cpf_cnpj: string;
  tipo: "pf" | "pj";
  rg?: string | null;
  data_nasc?: string | null;
  profissao?: string | null;
  renda?: string | null;
  telefone?: string | null;
  email?: string | null;
  endereco_json?: Record<string, string> | null;
  observacoes?: string | null;
}

export async function listClients(params?: { q?: string; cursor?: string }): Promise<ClientListPage> {
  const { data } = await api.get<ClientListPage>("/v1/clients", { params });
  return data;
}

export async function createClient(body: ClientIn): Promise<ClientOut> {
  const { data } = await api.post<ClientOut>("/v1/clients", body);
  return data;
}

export async function getClient(id: string): Promise<ClientOut> {
  const { data } = await api.get<ClientOut>(`/v1/clients/${id}`);
  return data;
}

export async function updateClient(id: string, body: ClientIn): Promise<ClientOut> {
  const { data } = await api.patch<ClientOut>(`/v1/clients/${id}`, body);
  return data;
}

export async function deactivateClient(id: string): Promise<ClientOut> {
  const { data } = await api.post<ClientOut>(`/v1/clients/${id}/deactivate`);
  return data;
}
